python/cudf/cudf/core/dataframe.py

```diff
@@ -2633,6 +2633,71 @@ def insert(self, loc, name, value):
 
         self._data.insert(name, value, loc=loc)
 
+    def diff(self, periods=1, axis=0):
+        """
+        Calculates the difference of a Dataframe element compared with
+        another element in the Dataframe, treating each column independently
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Periods to shift for calculating difference,
+            accepts negative values.
+        axis : {0 or ‘index’, 1 or ‘columns’}, default 0
+            Take difference over rows (0) or columns (1).
+            Only row-wise (0) shift is supported.
+
+        Returns
+        -------
+        DataFrame
+            First differences of the DataFrame.
+
+        Notes
+        -----
+        Diff currently only supports float and integer dtype columns with
+        no null values.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> gdf = cudf.DataFrame({'a': [1, 2, 3, 4, 5, 6],
+                                  'b': [1, 1, 2, 3, 5, 8],
+                                  'c': [1, 4, 9, 16, 25, 36]})
+        >>> gdf
+        a  b   c
+        0  1  1   1
+        1  2  1   4
+        2  3  2   9
+        3  4  3  16
+        4  5  5  25
+        5  6  8  36
+        >>> gdf.diff(periods=2)
+            a     b     c
+        0  <NA>  <NA>  <NA>
+        1  <NA>  <NA>  <NA>
+        2     2     1     8
+        3     2     2    12
+        4     2     3    16
+        5     2     5    20
+
+        """
+
+        if not axis == 0:
+            raise NotImplementedError("Only axis=0 is supported.")
+
+        if self.isna().any().any():
```

Comment:
Why is this constraint added?

Reply:
df.diff() currently only supports float and integer dtype columns with no null values.
- This constraint checks whether the dataframe contains any NaN values

Reply:
Why is that having NaN values unsupported?

Reply:
I was basing it on the current implem. for [cudf.Series.diff() ](https://github.com/rapidsai/cudf/blob/2ecebe17cf519db1bc7aa8d459a735ea9f90c199/python/cudf/cudf/core/series.py#L3435). Unsure if that is the correct approach or not.. feedback is appreciated 👍🏾

Reply:
The current column diff implementation isn't using libcudf. The age of the git blame suggests we wrote this before we had clean null handling in our numba.cuda utilities (3 years ago)

Reply:
In any case I think if the issue is with `diff`, then we should just let `diff` be the thing that throws the error (it appears to have a guard already anyways)

Reply:
`diff  = self - self.shift(periods)` should work for nulls too.

`shift` uses libcudf function. 
`-` uses binops libcudf function.
both supports nulls.

If this should be even optimized (without temporary shifted column and do it in single op), writing libcudf function is one option.

Reply:
Update: I removed the NaN - constraint check. Ready for review. Thanks @karthikeyann